packages/@lwc/style-compiler/src/serialize.ts

```diff
@@ -174,34 +174,46 @@ function serializeCss(result: Result, collectVarFunctions: boolean): string {
     return generateExpressionFromTokens(tokens);
 }
 
-// TODO [#1288]: this code needs refactor, it could be simpler by using a native post-css walker
-function tokenizeCssSelector(data: string): Token[] {
+// Given any CSS string, replace the scope tokens from the CSS with code to properly
+// replace it in the stylesheet function.
+function tokenizeCss(data: string): Token[] {
     data = data.replace(/( {2,})/gm, ' '); // remove when there are more than two spaces
-    const tokens: Token[] = [];
-    let pos = 0;
-    let next = 0;
-    const max = data.length;
 
-    while (pos < max) {
-        if (data.indexOf(`[${HOST_ATTRIBUTE}]`, pos) === pos) {
-            tokens.push({ type: TokenType.identifier, value: HOST_SELECTOR_IDENTIFIER });
+    const tokens: Token[] = [];
+    const regex = new RegExp(`[[-](${SHADOW_ATTRIBUTE}|${HOST_ATTRIBUTE})]?`, 'g');
 
-            next += HOST_ATTRIBUTE.length + 2;
-        } else if (data.indexOf(`[${SHADOW_ATTRIBUTE}]`, pos) === pos) {
-            tokens.push({ type: TokenType.identifier, value: SHADOW_SELECTOR_IDENTIFIER });
+    let lastIndex = 0;
+    for (const match of matchAll(data, regex)) {
+        const index = match.index!;
+        const [matchString, substring] = match;
 
-            next += SHADOW_ATTRIBUTE.length + 2;
-        } else {
-            next += 1;
+        if (index > lastIndex) {
+            tokens.push({ type: TokenType.text, value: data.substring(lastIndex, index) });
+        }
 
-            while (data.charAt(next) !== '[' && next < max) {
-                next++;
-            }
+        const identifier =
+            substring === SHADOW_ATTRIBUTE ? SHADOW_SELECTOR_IDENTIFIER : HOST_SELECTOR_IDENTIFIER;
 
-            tokens.push({ type: TokenType.text, value: data.slice(pos, next) });
+        if (matchString.startsWith('[')) {
+            // attribute in a selector, e.g. `foo[__shadowAttribute__]`
+            tokens.push({
+                type: TokenType.identifier,
+                value: identifier,
+            });
+        } else {
+            // suffix for an at-rule, e.g. `@keyframes spin-__shadowAttribute__`
+            tokens.push({
+                type: TokenType.expression,
+                // remove the '[' and ']' at the beginning and end, add an initial '-'
+                value: `${identifier} ? ('-' + ${identifier}.substring(1, ${identifier}.length - 1)) : ''`,
```

Comment:
This is a bit gnarly, but I think it's better to make keyframe/animation scoping complicated rather than attribute scoping. The reason is that one is much rarer than the other.

E.g. we could change it so that the string passed in to the `stylesheet` function is `"x-foo_foo"` rather than `"[x-foo_foo]"`, but then we would need to add `"[" + identifier + "]"`. Plus, we would need to handle the case where the identifier is an empty string – in which case we'd need to exclude the brackets, because we want selectors like `div` and not `div[]`.

Reply:
This expression can be hoisted at the root of the stylesheet function. At worst the variable will be tree-shaken if it is not used in the stylesheet.

```js
function stylesheet(hostSelector, shadowSelector, nativeShadow) {
  const keyframeShadowSelector = shadowSelector ? ('-' + shadowSelector.slice(1, -1)) : '';
  // ...
}
```

> E.g. we could change it so that the string passed in to the `stylesheet` function is `"x-foo_foo"` rather than `"[x-foo_foo]"`, but then we would need to add `"[" + identifier + "]"`.

We can also apply a similar approach hoisting approach. We can do this in a follow-up PR.

Reply:
Actually it seems that's not the case. I guess `terser` doesn't tree-shake it because of the `.slice()`:

```js
// tmp.js
function foo () {
  const shadowSelector = 'foo'
  const keyframeShadowSelector = shadowSelector ? ('-' + shadowSelector.slice(1, -1)) : '';
  console.log(shadowSelector)
}
foo()
```

```js
// tmp.min.js (`terser -mc`)
function foo(){const o="foo";o.slice(1,-1);console.log(o)}foo();
```